### DISCUSSION TITLE
can i use materialFromPoint in webXr ar mode ?

### DISCUSSION THREAD

**Discussion by Hamidrzash** [google/model-viewer#4847] [Q&A] [answered]
my web App is almost complete and the only issue I'm facing now is that this method is not getting called when im in webxr ar

const changeColorClick = async (event) => {
            console.log('getMaterialFromPoint')
            const material = modelViewer.materialFromPoint(event.clientX, event.clientY);
        };
        
modelViewer.addEventListener("touch", changeColorClick);

i dont know if i can get it working or not in ar😅am i listening to the wrong event maybe?

**Comment by elalish** [comment]
I don't think there's a "touch" event on the web - perhaps you mean "touchstart"? https://developer.mozilla.org/en-US/docs/Web/API/Touch_events

**Reply by Hamidrzash** [reply]
yes sorry i meant touchstart, but touchstart or touchend also dosent seem to work in webxr ar for me
modelViewer.addEventListener("touchstart", changeColorClick);

**Reply by elalish** [reply]
Hmm, I haven't actually tried this in WebXR mode. Have you tried listening for touches on the whole `<body>`? Or perhaps you only get this: https://developer.mozilla.org/en-US/docs/Web/API/XRSession/select_event. If that's true, we might need to re-emit the event from MV to make it accessible. Would you like to try poking around a bit? PRs welcome, as well as any details you learn.

**Reply by Hamidrzash** [reply]
Thanks for your answer. I will try out different ways. How can I listen to the select event?

**Reply by Hamidrzash** [reply]
My issue is now fully fixed by going back to version 3.4.0 and listening to click event on document (or window or html)
Latest version (3.5.0) also works with this but the result of materialFromPoint is not satisfying and wrong sometimes (just in webXR).
Also in the pickMaterial [example](https://modelviewer.dev/examples/scenegraph/#pickMaterialExample) documents its stated that it will work with modelViewer.addEventListener but its not working on any modelViewer version that I tried;

**Reply by elalish** [reply]
>version 1.4.0 and listening to click event on document (or window or html) Latest version (1.5.0) also works

Is this a typo? Our latest version is v3.5, and v1.4 is *very* old.

**Reply by Hamidrzash** [reply]
Yess sorry again I meant 3. not 1. in versions.

**Reply by elalish** [reply]
Interesting - I don't think any changes were made to `materialFromPoint` between v3.4 and v3.5 - can you create a site where we can repro the issue? Maybe post a screen recording? Agreed that our material picking example doesn't actually work in WebXR mode - it probably should!

**Reply by Hamidrzash** [reply]
Sure here are the links

https://chair.faramaham.com/modelviewer_3.4.0.html
https://chair.faramaham.com/modelviewer_3.5.0.html

Yes I also explored the changes made from 3.4.0 to 3.5.0 myself and couldn't find any issue or change that could cause this problem (a bug in the raycasting of the newer 'three.js' versions maybe?).

**Reply by elalish** [reply]
Indeed, you've found a significant regression! This isn't just `materialFromPoint` - all of the interactions are offset somehow. You can tell by touching the base of any model in WebXR mode - it shouldn't move until you slide your finger, but a tap is pulling it. So yes, something regressed, probably in three.js. Ugh...

**Reply by Hamidrzash** [reply]
Thanks for confirming the regression, @elalish

**Reply by elalish** [reply]
@mrdoob I'm at a loss - it's easy to reproduce the regression in WebXR mode with the above two links. Nothing really changed in MV between these releases; I went through the PRs. I also don't see anything that obviously changed in Three.js between 160 and 163. Yet somehow the raycaster and the renderer are not lining up, despite using the same camera and scene. I can't even figure out how to go about debugging this problem. Do you have any ideas?

**Reply by Hamidrzash** [reply]
I have been testing model-viewer-module with different 'three' versions and the problem seems to be between 160 and 161.
https://github.com/mrdoob/three.js/compare/r160...r161 
Thats 175 commits😄

**Reply by Hamidrzash** [reply]
I've identified that the issue seems to be from these lines added in the last [commit](https://github.com/mrdoob/three.js/pull/27586/commits/f0b6a7c0eec98ef5a3d6be50b8a23b4a58550f20) for version 161 of three.js
```
cameraL.near = _nearOverride;
cameraL.far = _farOverride;
cameraR.near = _nearOverride;
cameraR.far = _farOverride;

cameraL.updateProjectionMatrix();
cameraR.updateProjectionMatrix();
camera.updateProjectionMatrix();
```
The problem appears to be related to the `cameraL.updateProjectionMatrix();` call, which affects the subsequent code:
```
if ( cameras.length === 2 ) {
				setProjectionFromUnion( cameraXR, cameraL, cameraR );
			} else {
				// assume single camera setup (AR)
				cameraXR.projectionMatrix.copy( cameraL.projectionMatrix );
			}
```
I'm not entirely sure how this can be fixed yet.

**Reply by elalish** [reply]
Wow, good sleuthing! Have you filed a bug on Three.js yet?

**Reply by elalish** [reply, marked as answer]
Looks like this was fixed by https://github.com/mrdoob/three.js/pull/29120 in r168 💯
